pkg/agent/client.go

```diff
@@ -505,10 +544,16 @@ func (a *Client) remoteToProxy(connID int64, ctx *connContext) {
 			klog.ErrorS(err, "connection read failure", "connectionID", connID)
 			return
 		} else {
-			resp.Payload = &client.Packet_Data{Data: &client.Data{
-				Data:      buf[:n],
-				ConnectID: connID,
-			}}
+			data := make([]byte, 0, n)
```

Comment:
This decoupling is based on the fact we can store/cache the data received from the remote before sending it to the server. 
If the connection to the server is unhealthy or closed, maybe it's in some sense a good idea to block receiving more data from remote until the healthy connection to the server is back, because the cached data won't be sent to the server for connection is closed otherwise.

Reply:
That's a really nice idea for an improvement. However it would require we implement a resume all the way through. It involved being able to move in flight connections from 1 channel to newer channel. That means we need to coordinate that in both agent and server. While I really like this idea I think its better done in a subsequent change.